src/main/asciidoc/_chapters/offheap_read_write.adoc

```diff
@@ -0,0 +1,146 @@
+we re-desgined the read path and made the HDFS block reading be offheap now. This feature will be available in HBASE-3.0.0.
+
+For more details about the design and performance improvement, please see the link:https://docs.google.com/document/d/1xSy9axGxafoH-Qc17zbD2Bd--rWjjI00xTWQZ8ZwI_E/edit?usp=sharing[Document].
+Here we will share some best practice about the performance tuning:
+
+Firstly,  we introduced several configurations about the ByteBuffAllocator (which was abstracted to manage the memory application or release):
+
+1. hbase.ipc.server.reservoir.minimal.allocating.size: If the desired byte size is not less than this one, then it will be allocated as a pooled offheap ByteBuff, otherwise it will be allocated from heap directly because it
+is too wasting to allocate from pool with fixed-size ByteBuffers, default value is hbase.ipc.server.allocator.buffer.size/6.
+2. hbase.ipc.server.allocator.max.buffer.count: The ByteBuffAllocator will have many fixed-size ByteBuffers inside which are composited as a pool, this config indicate how many buffers are there in the pool.
+3. hbase.ipc.server.allocator.buffer.size: The byte size of each ByteBuffer, default value is 66560 (65KB).
+
+Second, we have some suggestions:
+
+.Please make sure that there are enough pooled DirectByteBuffer in your ByteBuffAllocator.
+
+The ByteBuffAllocator will allocate ByteBuffer from DirectByteBuffer pool firstly, if there’s no available ByteBuffer
+from the pool,  then it will just allocate the ByteBuffers from heap, then the GC pressures will increase again.
+
+By default, we will pre-allocate 2MB for each RPC handlers ( The handler count is determined by the config:
+hbase.regionserver.handler.count, it has the default value 30) . That’s to say,  if your hbase.ipc.server.allocator.buffer.size
+is 65KB, then your pool will have 2MB / 65KB * 30 = 945 DirectByteBuffer.  If you have some large scan and have a big caching,
+say you may have a rpc response whose bytes size is greater than 1MB (another 1MB for receiving rpc request),  then it will
+be better to increase the hbase.ipc.server.allocator.max.buffer.count.
+
+The RegionServer web UI also has the statistic about ByteBuffAllocator:
+
+If the following condition meet, you may need to increase your max buffer.count:
+
+image::bytebuff-allocator-stats.png[]
+
+heapAllocationRatio >= hbase.ipc.server.reservoir.minimal.allocating.size / hbase.ipc.server.allocator.buffer.size * 100%
+
+.Please make sure the buffer size is greater than your block size.
+
+We have the default block size=64KB, so almost all of the data block have a block size: 64KB + delta, whose delta is
+very small, depends on the size of last KeyValue. If we use the default hbase.ipc.server.allocator.buffer.size=64KB,
+then each block will be allocated as a MultiByteBuff:  one 64KB DirectByteBuffer and one HeapByteBuffer with delta bytes,
```

Comment:
Avoid referring as MBB which is internal stuff.  Say 2 BBs will be required then to read and carry the reading HFile block

Reply:
OK